### PR TITLE
Userland: Add {md5,sha1,sha256,sha512}sum

### PR DESCRIPTION
**Pull request by linusg**
~The program code is shared by using a huge macro that's just being passed the name of the hashing algorithm - hope that's a decent solution here?!~ (not anymore) I certainly don't want to maintain four separate copies of basically the same program.

Other than that I mimicked the behaviour of those programs on my Linux machine, for now they support multiple files, reading from stdin, file paths or both.

image

*Not sure what that setpgid message means...*

### CODE REVIEW

**Comment by alimpfard** [comment]
> Not sure what that setpgid message means...

Probably Shell bug lul

**Comment by alimpfard** [comment]
You could also make them all the same binary that chooses between the calls by the name it was invoked as

**Comment by linusg** [comment]
> You could also make them all the same binary that chooses between the calls by the name it was invoked as

yeah that was my first idea too, but then you have to have a common binary around that's all of these programs but also none of them at the same time and that's kind of awkward?!

**Comment by asynts** [comment]
> You could also make them all the same binary that chooses between the calls by the name it was invoked as.

Serenity has `alias` right? Just set `alias sha256sum='checksum --sha256'`.

**Comment by alimpfard** [comment]
> that's kind of awkward

busybox: \*cries in awkward\*

**Comment by linusg** [comment]
aaaaand I forgot a `#pragma once`.

**Comment by alimpfard** [comment]
> `alias sha256sum='checksum --sha256'`

That's a better solution, yeah!

**Comment by awesomekling** [comment]
Aliases won't work nicely in some situations though, e.g if you're running an autoconf-generated configure script via bash inside Serenity.

**Comment by alimpfard** [comment]
So...how about a `checksum` program, and individual `foosum` programs just `execv`'ing them with `--foo`?

**Comment by nico** [comment]
I think "checksum" binary that changes behavior based on its name + symlinks is nice and reasonably idiomatic (clang/clang++/clang-cl do this too for example).

**Comment by linusg** [comment]
Done.

File: Userland/checksum.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    Vector<const char*> paths;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(paths, paths_help_string.characters(), "path", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    if (paths.is_empty())
+        paths.append("-");
+
+    Crypto::Hash::Manager hash;
```
Comment (alimpfard) [inline Userland/checksum.cpp:69]: I see you found this (badly named...) guy.
I was gonna suggest it lol

**Comment by awesomekling** [comment]
Very nice!

**Pull request merged**